src/cli/mod.rs

```diff
@@ -0,0 +1,69 @@
+mod messages;
+
+use std::io::BufRead;
+use std::io::Write;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn get_user_input_works() {
+        let input = b"12345 65432";
+
+        let captured = get_user_input(&input[..]);
+        assert_eq!(String::from("12345 65432"), captured, "Expected input to be captured, was not");
+    }
+
+    #[test]
+    fn prompt_prints_prompt_to_output() {
+        let mut output = Vec::new();
+
+        prompt_for_user_input(&mut output);
+
+        let output = String::from_utf8(output).expect("Expected UTF-8 output, was not");
+
+        assert!(output.contains("Dice"),
+            "Expected output to contain prompt for dice throws, did not");
+    }
+
+    #[test]
+    fn welcome_prints_basic_guidance_to_output() {
+        let mut output = Vec::new();
+
+        welcome(&mut output);
+
+        let output = String::from_utf8(output).expect("Expected UTF-8 output, was not");
+
+        assert!(output.contains("https://www.eff.org/dice"),
+            "Expected output to contain guidance on generating passphrases using dice, did not");
+    }
+
+    #[test]
+    fn welcome_prints_version_number_to_output() {
+        let mut output = Vec::new();
+
+        welcome(&mut output);
+
+        let output = String::from_utf8(output).expect("Expected UTF-8 output, was not");
+
+        assert!(output.contains("v1.0.0-alpha"),
+            "Expected output to contain Dice version number, did not");
+    }
+}
+
+pub fn get_user_input<R>(mut input: R) -> String where R: BufRead {
+    let mut throws = String::new();
+    input.read_line(&mut throws)
+        .expect("Failed to read line from input");
+    throws
+}
+
+pub fn prompt_for_user_input<W>(mut output: W,) where W: Write {
+    write!(&mut output, "{}", messages::PROMPT).expect("Unable to write to output");
+    output.flush().expect("Unable to write to output");
+}
+
+pub fn welcome<W>(mut output: W) where W: Write {
```

Comment:
I didn't investigate much, but I noticed that if I didn't use the `where W: Write` and replaced `W` by `Write` everywhere, then `use std::io::Write;` would be _unused_ and `output` would miss the `std::io::Write` trait.

```
error[E0599]: no method named `write_fmt` found for type `&mut Write` in the current scope
  --> src/cli/mod.rs:68:5
   |
68 |     write!(&mut output, "{}", messages::WELCOME).expect("Unable to write to output");
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the method `write_fmt` exists but the following trait bounds were not satisfied:
           `&mut Write : std::io::Write`
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `write_fmt`, perhaps you need to implement one of them:
           candidate #1: `std::io::Write`
           candidate #2: `std::fmt::Write`
```